src/cpp/client/channel_cc.cc

```diff
@@ -35,17 +35,151 @@
+      gpr_thd_options options = gpr_thd_options_default();
+      gpr_thd_options_set_joinable(&options);
+      gpr_thd_new(&thd_id_, &WatchStateChange, this, &options);
+    }
+  }
+
+  ~ChannelConnectivityWatcher() {
+    cq_.Shutdown();
+    if (thd_id_ != 0) {
+      gpr_thd_join(thd_id_);
+    }
+  }
+
+  void WatchStateChangeImpl() {
+    bool ok = false;
+    void* tag = NULL;
+    CompletionQueue::NextStatus status = CompletionQueue::GOT_EVENT;
+    while (status != CompletionQueue::SHUTDOWN) {
+      status = cq_.AsyncNext(&tag, &ok, gpr_inf_past(GPR_CLOCK_REALTIME));
+      // Make sure we've seen 2 TIMEOUTs before going to sleep
+      if (status == CompletionQueue::TIMEOUT) {
+        status = cq_.AsyncNext(&tag, &ok, gpr_inf_past(GPR_CLOCK_REALTIME));
+      }
+      if (status == CompletionQueue::TIMEOUT) {
+        gpr_sleep_until(
+            gpr_time_add(gpr_now(GPR_CLOCK_REALTIME),
+                         gpr_time_from_millis(kConnectivityCheckIntervalMsec,
+                                              GPR_TIMESPAN)));
+      } else if (status == CompletionQueue::GOT_EVENT) {
+        ChannelState* channel_state = static_cast<ChannelState*>(tag);
+        channel_state->state = grpc_channel_check_connectivity_state(
+            channel_state->channel, false);
+        if (channel_state->state == GRPC_CHANNEL_SHUTDOWN) {
+          void* shutdown_tag = NULL;
+          channel_state->shutdown_cq.Next(&shutdown_tag, &ok);
+          delete channel_state;
+        } else {
+          TagSaver* tag_saver = new TagSaver(channel_state);
+          grpc_channel_watch_connectivity_state(
+              channel_state->channel, channel_state->state,
+              gpr_inf_future(GPR_CLOCK_REALTIME), cq_.cq(), tag_saver);
+        }
+      }
+    }
+  }
+
+  void StartWatching(grpc_channel* channel) {
+    if (thd_id_ != 0) {
+      ChannelState* channel_state = new ChannelState(channel);
+      // The first grpc_channel_watch_connectivity_state() is not used to
+      // monitor the channel state change, but to hold a reference of the
+      // c channel. So that WatchStateChangeImpl() can observe state ==
+      // GRPC_CHANNEL_SHUTDOWN before the channel gets destroyed.
+      grpc_channel_watch_connectivity_state(
+          channel_state->channel, channel_state->state,
+          gpr_inf_future(GPR_CLOCK_REALTIME), channel_state->shutdown_cq.cq(),
+          new TagSaver(nullptr));
+      grpc_channel_watch_connectivity_state(
+          channel_state->channel, channel_state->state,
+          gpr_inf_future(GPR_CLOCK_REALTIME), cq_.cq(),
+          new TagSaver(channel_state));
+    }
+  }
+
+ private:
+  struct ChannelState {
+    explicit ChannelState(grpc_channel* channel)
+        : channel(channel), state(GRPC_CHANNEL_IDLE){};
+    grpc_channel* channel;
+    grpc_connectivity_state state;
+    CompletionQueue shutdown_cq;
+  };
+  gpr_thd_id thd_id_;
+  CompletionQueue cq_;
+  CompletionQueue shutdown_cq_;
```

Comment:
It doesn't look like this is actually used anywhere.  The code on lines 119 and 140 uses `channel_state->shutdown_cq` instead.